### DISCUSSION TITLE
One AppHost instance monitoring multiple GRPC clients, how-to?

### DISCUSSION THREAD

**Discussion by adivoj** [microsoft/aspire#3432] [Q&A] [answered]
Hi dear fellows,

I have multiple GRPC service clients as NuGet packages and I would like to have one solution (dashboard) that will have the traces of my clients. Is this possible? What should that standalone solution have and what should my clients have? 

Thanks :)

**Comment by davidfowl** [comment]
It's unclear what you mean here. What are you trying to do? Do you have an application that uses those clients? What have you tried to make work?

**Comment by adivoj** [comment]
Hi David,

I have multiple applications using multiple clients. Can it be unified under one dashboard?

I've tried this in a new solution that runs the dashboard:
```
var builder = DistributedApplication.CreateBuilder(args);
builder.Services.AddGrpcClient<MyCompany.ClientNameGrpcServiceClient>("clientName", o =>
     {
         o.Address = new Uri("https://localhost:5105");
     })
     .AddCallCredentials((context, metadata) => ...
     .ConfigureChannel(o => ...

builder.Build().Run();
```

And this in the client:
```
var builder = WebApplication.CreateBuilder(args);
builder.AddServiceDefaults();
```

If I have Aspire.AppHost with a direct project reference in the client itself like below, then it works.. But I need a standalone dashboard tracing the communication between my applications and grprc clients.
```
var builder = DistributedApplication.CreateBuilder(args);
var apiService = builder
    .AddProject<Projects.Grpc_Server>("grpcserver")
    .WithLaunchProfile ("https");
builder.Build().Run();
```

**Comment by davidfowl** [comment]
> But I need a standalone dashboard tracing the communication between my applications and grprc clients.

This was a piece of missing information from the original question. So you're trying to figure out how to point multiple projects at the same standalone dashboard? What did you try and what's not working?

**Comment by adivoj** [comment]
Hi,

Yes.

I've tried doing it with above code. Code 1 in the standalone dashboard and Code 2 in the clients. Is this the way how to do it? I can open the dashboard but I don't see any resources.

**Reply by davidfowl** [reply]
Where did you configure the application to send telemetry to the dashboard?

**Reply by adivoj** [reply]
In code snippet 2:
```
var builder = WebApplication.CreateBuilder(args);
builder.AddServiceDefaults();
```

The AddServiceDefaults() adds the GRPC tracing. That's what I have in the grpc client and if I put the AspireApphost in the same solution with direct project references, I can see the traces.

**Reply by davidfowl** [reply]
That's because the apphost is injecting the OTLP endpoint into your application. When you run the dashboard in standalone mode, you have to do that work (as you have lost orchestration).

**Reply by adivoj** [reply]
Ok, thanks David! 
Do you have any links to docs where I can pick up some code?

**Reply by davidfowl** [reply, marked as answer]
The aspire docs are here https://learn.microsoft.com/en-us/dotnet/aspire/get-started/aspire-overview

There's a standalone dashboard sample here:

https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard#standalone-mode

**Reply by adivoj** [reply]
Oki, thanks David, that should get me going. I think I was missing **OTEL_EXPORTER_OTLP_ENDPOINT** in the appsettings.
